hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/compact/CompactionPlanSourceFunction.java

```diff
@@ -48,47 +60,154 @@
-  public void run(SourceContext sourceContext) throws Exception {
-    List<CompactionOperation> operations = this.compactionPlan.getOperations().stream()
+  public void run(SourceContext<CompactionPlanEvent> context) throws Exception {
+    if (isStreamingMode) {
+      while (isRunning) {
+        monitorCompactionPlan(context);
+        TimeUnit.SECONDS.sleep(interval);
+      }
+    } else {
+      monitorCompactionPlan(context);
+    }
+  }
+
+  public void monitorCompactionPlan(SourceContext<CompactionPlanEvent> context) throws IOException {
+    table.getMetaClient().reloadActiveTimeline();
+
+    // checks the compaction plan and do compaction.
+    if (OptionsResolver.needsScheduleCompaction(conf)) {
+      Option<String> compactionInstantTimeOption = CompactionUtil.getCompactionInstantTime(table.getMetaClient());
+      if (compactionInstantTimeOption.isPresent()) {
+        boolean scheduled = writeClient.scheduleCompactionAtInstant(compactionInstantTimeOption.get(), Option.empty());
+        if (!scheduled) {
+          // do nothing.
+          LOG.info("No compaction plan for this job ");
+          return;
+        }
+        table.getMetaClient().reloadActiveTimeline();
+      }
+    }
+
+    // fetch the instant based on the configured execution sequence
+    String compactionSeq = conf.getString(FlinkOptions.COMPACTION_SEQUENCE);
+    HoodieTimeline timeline = table.getActiveTimeline().filterPendingCompactionTimeline();
+    Option<HoodieInstant> requested = CompactionUtil.isLIFO(compactionSeq) ? timeline.lastInstant() : timeline.firstInstant();
+    if (!requested.isPresent()) {
+      // do nothing.
+      LOG.info("No compaction plan scheduled, turns on the compaction plan schedule with --schedule option");
+      return;
+    }
+
+    String compactionInstantTime = requested.get().getTimestamp();
+
+    HoodieInstant inflightInstant = HoodieTimeline.getCompactionInflightInstant(compactionInstantTime);
+    if (timeline.containsInstant(inflightInstant)) {
+      LOG.info("Rollback inflight compaction instant: [" + compactionInstantTime + "]");
+      table.rollbackInflightCompaction(inflightInstant);
+      table.getMetaClient().reloadActiveTimeline();
+    }
+
+    // generate compaction plan
+    // should support configurable commit metadata
+    HoodieCompactionPlan compactionPlan = CompactionUtils.getCompactionPlan(
+        table.getMetaClient(), compactionInstantTime);
+
+    if (compactionPlan == null || (compactionPlan.getOperations() == null)
+        || (compactionPlan.getOperations().isEmpty())) {
+      // No compaction plan, do nothing and return.
+      LOG.info("No compaction plan for instant " + compactionInstantTime);
+      return;
+    }
+
+    HoodieInstant instant = HoodieTimeline.getCompactionRequestedInstant(compactionInstantTime);
+    HoodieTimeline pendingCompactionTimeline = table.getActiveTimeline().filterPendingCompactionTimeline();
+    if (!pendingCompactionTimeline.containsInstant(instant)) {
```

Comment:
Is it necessary to start a separate scheduling for compaction ? There is only possible compaction plans when an instant is committed successfully, seems more reasonable to schedule the compaction on ckp success event, based on the fact users may have custom ckp interval config.